### ISSUE TITLE
CheckEvenEqn assertion failure triggered via library interface

### ISSUE THREAD

**Issue by smcc** [stp/stp#162]
I've found an interesting assertion failure which is caused by a query expressed with the library interface, though not triggered by the analogous inputs my tool can produce using other interfaces. I was able to extract a test case using the fprintf+perl script I'd mentioned in another report:

https://github.com/bitblaze-fuzzball/stp/compare/bitblaze-fuzzball:master...c_interface-trace

Along with some manual editing and simplification:

https://gist.github.com/smcc/90717a5048a50527641d

The simplified version isn't simple enough to make the problem obvious, but I hope it's at least short enough to be easy to work with.

The assertion is in BVSolver::CheckEvenEqn, a method which is part of a specialized algorithm for simplifying linear bitvector equations in which all coefficients and the constant term are even. Part of the code expects that the constant term will appear on one side of the equality and all the terms on the other side will be non-constant, though elsewhere it also seems to only handle the case where the constant is zero. But if it happens to find a constant term on the addition side of the equation (before finding any other terms that are not the product of an even value times a symbol), this assertion failure is triggered. The comment on the assertion says that the algorithm would work incorrectly in this case, and I can see that: it doesn't have any support for adding multiple constants, so for instance if one was odd that doesn't mean the same thing as if the sum of the contstant terms was odd.

Since this entire part of the system is an optimization, and this case clearly doesn't happen often, it seems like the easiest fix would be to have the code just report in this case that the solving approach is not applicable, as for instance it would if there was a non-linear term in the equation, by setting the "evenflag" variable to false and returning its input unchanged. Though perhaps my test case indicates another problem where earlier in processing an equation is not getting simplified in the way the rest of the code expects it to be.

**Comment by msoos** [comment]
We are stuck here:
```
Input to BVSolve_Odd()
602:(EQ 
  34:0x00000000
  600:(BVPLUS 
    40:0x00000044
    586:0xFFFFFFBC))

AT node:
40:0x00000044
 (kind: BVCONST )
AT node:
586:0xFFFFFFBC
 (kind: BVCONST )
```

And as you note, that's 2 constants in a row. I'm wondering if that was supposed to have been eliminated by systems that collapse constants before this function is called. There are extensive systems that detect and deal with constants. I'm looking into this.

**Comment by msoos** [comment]
I solved it the simple way. If I see 2 or more constants, I simply exit with "unknown" leaving the system as it is. I believe this is a working solution.

As for it being the best solution or not is a different question. I think we could check the evenness of all the constants and if all of them are even then of course their sum will also be even. I think that would work. This code is super-special-case. I added a TODO so we can get back here and fix it later if we have time.

I hope it's OK.

Mate
